src/haddock/gear/known_cns_errors.py

```diff
@@ -0,0 +1,173 @@
+        file_handle = gzip.open(cns_out_fpath, "rb")
+    else:
+        file_handle = open(cns_out_fpath, "rb")
+    try:
+        _find_cns_errors(file_handle, KNOWN_ERRORS, filepath=cns_out_fpath)
+    except KnownCNSError as err:
+        return err
+    else:
+        return None
+
+
+def _find_cns_errors(
+        file_handle: Union[gzip.GzipFile, BufferedReader],
+        known_errors: dict[str, str],
+        chunk_size: int = 4096,
+        filepath: FilePath = "",
+        ) -> None:
+    """Backward reading and detect first known CNS error in file.
+
+    Parameters
+    ----------
+    file_handle: Union[gzip.GzipFile, BufferedReader]
+        An opened file in read bytes mode.
+    known_errors : dict[str, str]
+        Dict of known errors and their hints
+    chunk_size : int, optional
+        Check size (in bytes) to read the file backwards, by default 4096
+    filepath : FilePath -> Union[str, Path]
+        Path to the cns.out file currently checked.
+
+    Raises
+    ------
+    KnownCNSError
+        An exception for known CNS errors, with its hint on how to solve it!
+    """
+    # Find file size
+    file_handle.seek(0, 2)
+    size = file_handle.tell()
+    buffer = b''
+    parsed_lines = 99999
+    for i in range(size - 1, -1, -chunk_size):
+        # Go to location in file
+        file_handle.seek(max(i - chunk_size, 0))
+        # Read next chunk
+        chunk = file_handle.read(min(chunk_size, i + 1))
+        # Increment buffer
+        buffer = chunk + buffer
+        lines = buffer.split(b'\n')
+        # Read lines
+        for line in reversed(lines[-len(lines):parsed_lines]):
+            decoded_line = line.decode('utf-8', errors='replace')
+            # Loop over known errors
+            for error_string, hint in known_errors.items():
+                # Check if this error is known
+                if error_string in decoded_line:
+                    # return the cause
+                    raise KnownCNSError(
+                        error_string,
+                        hint,
+                        filepath,
+                        )
+        # Update number of parsed lines so we do not check them again
+        parsed_lines = -len(lines)
+
+
+def find_all_cns_errors(
+        directory_path: FilePath,
+        ) -> dict[str, dict[str, Union[int, KnownCNSError]]]:
+    """Find all errors in a directory.
+
+    Parameters
+    ----------
+    directory_path : FilePath
+        Path to the directory to be checked.
+
+    Returns
+    -------
+    all_errors : dict[str, dict[str, Union[list[FilePath], KnownCNSError]]]
+        Dictionary containing all errors found in this directory.
+    """
+    all_errors: dict[str, dict[str, Union[int, KnownCNSError]]] = {}
+    # Gather list of all `.out` and `.out.gz` files present in directory
+    all_cns_out_files = list(Path(directory_path).glob("*.out.gz"))
+    all_cns_out_files += list(Path(directory_path).glob("*.out"))
```

Comment:
The logic of this function is correct but this is in the wrong place, it needs to go here: https://github.com/haddocking/haddock3/blob/a15dc5e7569d53421c861dcef4980149d6cc90d4/src/haddock/libs/libsubprocess.py#L170-L174

And either parse the file or the output coming out of the subprocess